[pybind] New Runtime pybind API

larryliu0820 opened the pull request:
Summary:
    Based on this proposal: https://docs.google.com/document/d/10Q4-pt97inQQtFf-FjjwhMaDXXCfk1zGy6V6EkygNUY/edit#heading=h.fcrpnrtb6cud
    
Historically our pybinding APIs are not following the same C++ modeling
(Program, Method etc) and hence it's hard to use and easy to hit
footguns - for example, if we load the program and return it from a
python method, it goes out of the scope and releases the memory.

This effort is to create Pybind APIs that resembles C++ objects so it's
less confusing to the users.

Add the following python classes:
* `Runtime`: a singleton object hosting methods like `load_program`.
  Returns a `Program` object when calling `load_program`. Also exposes
  the operator registry
* `Program`: each pte file should have one `Program` object. Most
  important method is `load_method` which returns a `Method` object. It
  has a property `method_names` where we can inspect what methods are
  inside this .pte file.
* `Method`: one object per method name in a given `Program`. Exposes
  `execute` which takes in pytree flattened torch tensors as input and
  return pytree flattened output. It also exposes `MethodMeta` for users
  to inspect more information regarding input/output of this method.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

dbort commented on runtime/test/test_runtime.py:
```diff
@@ -0,0 +1,81 @@
+class RuntimeTest(unittest.TestCase):
+    def test_smoke(self):
+        ep, inputs = create_program(ModuleAdd())
+        runtime = Runtime.get()
+
+        program = runtime.load_program(
```
Please add a test that get() really returns a singleton.

```suggestion
        runtime = Runtime.get()

        # Demonstrate that get() returns a singleton.
        runtime2 = Runtime.get()
        self.assertTrue(runtime is runtime2)

        program = runtime.load_program(
```

dbort commented on runtime/__init__.py:
```diff
@@ -0,0 +1,208 @@
+    et_runtime: Runtime = Runtime.get()
+    program: Program = et_runtime.load_program(
+        Path("/tmp/program.pte"),
+        config=LoadProgramConfig(verification="internal_consistency"),
```
This is inconsistent with the implemented API

dbort commented on runtime/__init__.py:
```diff
@@ -0,0 +1,208 @@
+    from pathlib import Path
+
+    import torch
+    from executorch.runtime import LoadProgramConfig, Runtime
```
LoadProgramConfig doesn't exist

dbort commented on runtime/__init__.py:
```diff
@@ -0,0 +1,208 @@
+
+class Method:
+    """An ExecuTorch method, loaded from a Program.
+    TODO: This class should be pybind to the C++ counterpart instead of hosting ExecuTorchModule.
```
This TODO is more of a comment than a documentation item: users reading the rendered pydoc shouldn't see this.

dbort reviewed (changes requested):
This looks great! Mostly small issues.

larryliu0820 commented on runtime/__init__.py:
```diff
@@ -0,0 +1,208 @@
+            ModuleNotFoundError: The prebuilt _portable_lib.so is not found.
+        """
+        try:
+            import executorch.extension.pybindings.portable_lib as legacy_module
```
Seems like it's better to fail at the top of the file

larryliu0820 replied:
Actually I'll try to catch an exception at the top of the file

dbort commented on runtime/__init__.py:
```diff
@@ -93,7 +97,11 @@ def __init__(self, module: ExecuTorchModule, data: bytes) -> None:
         self._data = data
         self._module = module
-        self._methods = defaultdict(str)
+        self._methods = {}
```
```suggestion
        self._methods: Dict[Method] = {}
```

dbort commented on runtime/__init__.py:
```diff
@@ -110,17 +118,16 @@ def load_method(self, name: str) -> Method:
     """The registry of operators that are available to the runtime.
-
-    Currently only supports printing out all registered operator names.
+    # TODO: Expose the kernel callables to Python.
```
This should just be a comment, not part of the user-visible docs.

dbort commented on runtime/__init__.py:
```diff
@@ -110,17 +118,16 @@ def load_method(self, name: str) -> Method:
 
-    def __init__(self) -> None:
-        pass
+    def __init__(self, module: ModuleType) -> None:
+        self._legacy_module = module
```
To be consistent with the rest of the file

```suggestion
    def __init__(self, legacy_module: ModuleType) -> None:
        self._legacy_module = legacy_module
```

dbort commented on runtime/__init__.py:
```diff
@@ -142,67 +149,59 @@ class Runtime:
-                test/validation data.
-            verification_config: The configuration for program verification.
+            data: The binary program data to load; typically PTE data.
+            verification: The configuration for program verification.
```
"configuration" doesn't mean much here.
```suggestion
            verification: The level of program verification to perform.
```

dbort commented on runtime/__init__.py:
```diff
@@ -0,0 +1,207 @@
+    This can be used to load the methods/models defined by the program.
+    """
+
+    def __init__(self, module: ExecuTorchModule, data: bytes) -> None:
```
Since data can be None when loading from a file
```suggestion
    def __init__(self, module: ExecuTorchModule, data: Optional[bytes]) -> None:
```

dbort commented on runtime/__init__.py:
```diff
@@ -0,0 +1,207 @@
+    def method_names(self) -> Sequence[str]:
+        return set(self._methods.keys())
+
+    def load_method(self, name: str) -> Method:
```
Since this can return None. The "Returns" section should also mention the None case.
```suggestion
    def load_method(self, name: str) -> Optional[Method]:
```

dbort commented on runtime/__init__.py:
```diff
@@ -0,0 +1,207 @@
+
+    @property
+    def operator_names(self) -> Sequence[str]:
+        """Gets the names of all registered operators.
+
+        Returns:
+            The names of all registered operators.
+        """
```
Fine as-is, but if the summary for a method repeats the Returns section like this, it's better to condense it into a single line. And since it's a property, it doesn't really get or return, it just is.

```suggestion
        """The names of all registered operators."""
```

dbort commented on runtime/__init__.py:
```diff
@@ -0,0 +1,207 @@
+    @staticmethod
+    @functools.lru_cache(maxsize=1)
+    def get() -> "Runtime":
+        """Gets the Runtime singleton.
+
+        Raises:
+            ModuleNotFoundError: if the prebuilt _portable_lib.so is not found.
+        """
```
I don't think the Raises situation can happen given the check at the top of this file

```suggestion
        """Gets the Runtime singleton."""
```

dbort reviewed (approved):
A few final minor things, but looks great!

larryliu0820 committed:
New Runtime pybind API (#6063)

Summary:
Based on this proposal: https://docs.google.com/document/d/10Q4-pt97inQQtFf-FjjwhMaDXXCfk1zGy6V6EkygNUY/edit#heading=h.fcrpnrtb6cud
    
Historically our pybinding APIs are not following the same C++ modeling
(Program, Method etc) and hence it's hard to use and easy to hit
footguns - for example, if we load the program and return it from a
python method, it goes out of the scope and releases the memory.

This effort is to create Pybind APIs that resembles C++ objects so it's
less confusing to the users.

Add the following python classes:
* `Runtime`: a singleton object hosting methods like `load_program`.
  Returns a `Program` object when calling `load_program`. Also exposes
  the operator registry
* `Program`: each pte file should have one `Program` object. Most
  important method is `load_method` which returns a `Method` object. It
  has a property `method_names` where we can inspect what methods are
  inside this .pte file.
* `Method`: one object per method name in a given `Program`. Exposes
  `execute` which takes in pytree flattened torch tensors as input and
  return pytree flattened output. It also exposes `MethodMeta` for users
  to inspect more information regarding input/output of this method.


Reviewed By: dbort

Differential Revision: D64132360

Pulled By: larryliu0820
